elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,250 @@
+
+We may also consider supporting a default resolver, in that case the `expression=encryption` resource would have an additional attribute `default-resolver` to specify which resolver is the default, the expression can then be simplified to `${ENC:ENCRYPTED_DATA}`.  In this case the `resolver` parameter on the `create-expression` resource could also become optional allowing the default to be used.
+
+We will not support the decryption of expressions other than the support within the management model to resolve existing expressions.
+
+As expressions are already widely in use with the application server there is a small possibility that users may have already defined expressions with a prefix of `ENC:` expecting it to be resolved as a system property.  The `expression=encryption` resource will also have a `prefix` attribute which will allow an alternative prefix to `ENC` to be specified. 
+
+==== Command Line Encryption
+
+Whilst the Elytron tool does contain a `credential-store` command the encryption of values is really a special operation so a new command `encryption` will be added to the tool.
+
+This command will duplicate the parameters used by the `credential-store` command to configure the location and initialisation of the underlying credential store, additionally it will take an argument to specify the alias from the credential store to load the `SecretKey` to use for encryption.
+
+Additional arguments to the tool will be used to specify the algorithm to use for encryption as well as any additional parameters requires such as a salt or iteration count.
+
+The new `encryption` command will support two actions: -
+
+ * `--encrypt` - To encrypt a value and return the result Base64 encoded.
+ * `--create-expression` - To encrypt a value and return it as a complete expression.
+ 
+ In both cases the value to be encrypted can be passed in with the action or if not specified the user will be prompted to enter the value.  In the case of `--create-expression` the user can also optionally specify the `resolver` that will be used to create the expression String. 
+
+The tool will not add any support for the decryption of expressions or their values.
+
+==== Base64 Representations
+
+Base64 representations will be used for both the export and import of the generated keys and for the expression value to be passed to the expression resolver - generally these details should be opaque to the end user however we may include the description within the documentation allowing for others to make use of the format.  We should consider that in the future alternative representations may be required either to support alternative key types or to support alternative representations so the representations will also be versioned.
+
+The general structure of the first representation will be: -
+
+    { 'E', 'L', 'Y', (VERSION), (TYPE), VALUE... }
```

Comment:
I have never seen this format (structure) before. Can you please explain to me (and maybe add a note too to the document) what it represents? An example would be best.

Reply:
I am not sure what else to add for this, really it is a Java / pseudocode representation of the population of an array.

Reply:
So this represents and array - that's what wasn't clear to me. Thanks!

Reply:
I will change the wording above to say this is an array.